docling_core/types/experimental/document.py

```diff
@@ -0,0 +1,1106 @@
+                return data
+            text = data["bbox"].get("token", "")
+            if not len(text):
+                text_cells = data.pop("text_cell_bboxes", None)
+                if text_cells:
+                    for el in text_cells:
+                        text += el["token"] + " "
+
+                text = text.strip()
+            data["text"] = text
+
+        return data
+
+
+class BaseTableData(BaseModel):  # TBD
+    """BaseTableData."""
+
+    table_cells: List[TableCell] = []
+    num_rows: int = 0
+    num_cols: int = 0
+
+    @computed_field  # type: ignore
+    @property
+    def grid(
+        self,
+    ) -> List[List[TableCell]]:
+        """grid."""
+        # Initialise empty table data grid (only empty cells)
+        table_data = [
+            [
+                TableCell(
+                    text="",
+                    start_row_offset_idx=i,
+                    end_row_offset_idx=i + 1,
+                    start_col_offset_idx=j,
+                    end_col_offset_idx=j + 1,
+                )
+                for j in range(self.num_cols)
+            ]
+            for i in range(self.num_rows)
+        ]
+
+        # Overwrite cells in table data for which there is actual cell content.
+        for cell in self.table_cells:
+            for i in range(
+                min(cell.start_row_offset_idx, self.num_rows),
+                min(cell.end_row_offset_idx, self.num_rows),
+            ):
+                for j in range(
+                    min(cell.start_col_offset_idx, self.num_cols),
+                    min(cell.end_col_offset_idx, self.num_cols),
+                ):
+                    table_data[i][j] = cell
+
+        return table_data
+
+
+class FileInfo(BaseModel):
+    """FileInfo."""
+
+    filename: str
+    document_hash: str
+
+
+class RefItem(BaseModel):
+    """RefItem."""
+
+    cref: str = Field(alias="$ref")
+
+    # This method makes RefItem compatible with DocItem
+    def get_ref(self):
+        """get_ref."""
+        return self
+
+    model_config = ConfigDict(
+        populate_by_name=True,
+    )
+
+    def resolve(self, doc: "DoclingDocument"):
+        """resolve."""
+        path_components = self.cref.split("/")
+        if len(path_components) > 2:
+            _, path, index_str = path_components
+        else:
+            _, path = path_components
+            index_str = None
+
+        if index_str:
+            index = int(index_str)
+            obj = doc.__getattribute__(path)[index]
+        else:
+            obj = doc.__getattribute__(path)
+
+        return obj
+
+
+class ImageRef(BaseModel):
+    """ImageRef."""
+
+    format: str  # png, etc.
+    dpi: int  # ...
+    size: Size
+    uri: AnyUrl
+
+
+class ProvenanceItem(BaseModel):
+    """ProvenanceItem."""
+
+    page_no: int
+    bbox: BoundingBox
+    charspan: Tuple[int, int]
+
+
+class NodeItem(BaseModel):
+    """NodeItem."""
+
+    dloc: str  # format spec ({document_hash}{json-path})
```

Comment:
@ceberam Agreed, consider it open to edit.

Reply:
Since each node is already connected with its `parent` and `children`, I am wondering why we need to materialize the path of `dloc` (even including the actual doc ID/hash).

In which scenario is `dloc` going to be used?

Reply:
`dloc` is a unique identifier string for any element, and makes it traceable back to a field in a document. For example it can act as provenance in a vector store or other places.

Reply:
I see. The way I mean it is that materializing this info on the node level may be making the whole design a bit too rigid (for example the design would otherwise still allow e.g. easy removals or additions of nodes at random positions)

> it can act as provenance in a vector store

This could be under the responsibility of a chunker having to chunkify a whole doc to nodes, producing the paths while iterating.